nautobot/docs/development/best-practices.md

```diff
@@ -63,3 +63,193 @@ class ExampleModel(PrimaryModel):
+- Use of [`filter_overrides`](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#filter-overrides) **must be considered** in cases where local overrides to certain model field types are required vs. overloading or re-declaring the filter fields are required, such as changing a filter class, or customizing a UI widget. For example:
+
+```python
+class ProductFilter(NautobotFilterSet):
+
+     class Meta:
+         model = Product
+         fields = "__all__"
+         filter_overrides = {
+             models.CharField: {
+                 'filter_class': django_filters.CharFilter,
+                 'extra': lambda f: {
+                     'lookup_expr': 'icontains',
+                 },
+             },
+             models.BooleanField: {
+                 'filter_class': django_filters.BooleanFilter,
+                 'extra': lambda f: {
+                     'widget': forms.CheckboxInput,
+                 },
+             },
+         }
+```
+
+!!!warning
+    Existing features of filtersets and filters **must** be exhausted first using keyword arguments before resorting to customizing, re-declaring/overloading, or defining filter methods.
+
+### Filter Methods
+
+Filter methods in the current Nautobot core are problematic because they break the ability for such filter fields to be properly reversible. Specifically a filter method is a callable or the name of a method on a filterset that can be declared on a filter when it is defined on a filterset. This method is used to perform custom business logic on a filter field.
+
+Consider this example from `nautobot.dcim.filters.DeviceFilterSet`:
+
+```python
+  # Filter field definition is a BooleanFilter, for which an "isnull" lookup_expr 
+  # is the only valid filter expression
+  console_ports = django_filters.BooleanFilter(
+      method="_console_ports",
+      label="Has console ports",
+  )
+  
+  # Method definition loses context and further the field's lookup_expr 
+  # falls back to the default of "exact".
+  def _console_ports(self, queryset, name, value):
+      breakpoint()  # This is where we'll illustrate pdb below
+      return queryset.exclude(consoleports__isnull=value)
+```
+
+The default `lookup_expr` unless otherwise specified is “exact”, as seen in [django_filters.conf](https://github.com/carltongibson/django-filter/blob/main/django_filters/conf.py#L10):
+
+```python
+  'DEFAULT_LOOKUP_EXPR': 'exact',
+```
+
+When this method is called, the internal state is default, making reverse introspection impossible, because the `lookup_expr` is defaulting to “exact”:
+
+```python
+(Pdb) field = self.filters["console_ports"]
+(Pdb) field.exclude
+False
+(Pdb) field.lookup_expr
+'exact' 
+```
```

Comment:
Is this really an example of why `method`s are bad, or just an example of how core is not currently following all of the requirements laid out above regarding the correct use of `lookup_expr` and `exclude=True`? In other words, if we had instead:

```python
  console_ports = django_filters.BooleanFilter(
      method="_console_ports",
      label="Has console ports",
      exclude=True,
      lookup_expr="isnull",
  )
```

the method would still be *redundant* (as laid out below), but would it still be *a problem* as far as introspection and reversibility?

Or to put it yet another way, do we have any better examples where we have filter fields that are properly defined per all of the above requirements, but still need to use `method`, and that usage is still a problem?

Reply:
Solid point. I think we could refresh this to use a different field. I'll come up with one.

Reply:
To answer your questions, however:

> the method would still be redundant (as laid out below), but would it still be a problem as far as introspection and reversibility?

Redundant, yes. Problematic, maybe, since it provides context on how to proceed, if it's doing exactly the same logic as what the method would be doing, then it's not a problem. It becomes more problematic in the case where we don't actually know *what* the method is doing internally, and for example if it's generating a `Q` object internally. As developers/authors, that's not bad, as platfors maintainers trying to build reversible lookups, that's bad.

Reply:
For the sake of this thread, I would argue that `DeviceFilterSet.pass_through_ports` is bad. 

Here's why:

1. It's a method filter:

```python
    pass_through_ports = django_filters.BooleanFilter(
        method="_pass_through_ports",
        label="Has pass-through ports",
    )
```

2. It's generating a compound exclude filter using two related fields:

```python
    # 2.0 TODO: Remove me and `pass_through_ports` in exchange for `has_(front|rear)_ports`.
    def _pass_through_ports(self, queryset, name, value):
        return queryset.exclude(frontports__isnull=value, rearports__isnull=value)
```

3. `field_name` is irrelevant here because there are two at play (`frontports` and `rearports`)

4. There is no way to introspect what is going on in the method body with 

So while this filter definition could be improved like so:

```python
    pass_through_ports = django_filters.BooleanFilter(
        method="_pass_through_ports",
        label="Has pass-through ports",
        lookup_expr="isnull",
        exclude=True,
    )
```

It stops there.

- There's no way to identify either of the field names required here
- The `name` that is incoming to the method is the filter name as defined `has_pass_through_ports` in this case
- So the filter can be introspected for `lookup_expr` value using `self.filters[name].lookup_expr`, but it would have to be assumed that applies to both.
- Same with `exclude` (`self.filters[name].exclude`)

It would be better to just eliminate `pass_through_ports=True` entirely in exchange for `front_ports=True&rear_ports=True`  (current) or `has_frontports=True&has_rearports=True` (future)